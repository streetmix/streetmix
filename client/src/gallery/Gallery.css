body.gallery-visible {
  .main-screen {
    transform: translateY(calc(var(--gallery-height) - 90px));
  }

  .street-nameplate-container {
    transform: translateY(60px);
  }
}

body:not(.safari).gallery-visible .street-section-sky {
  transform: translateY(-20px);
}

.gallery-panel {
  --gallery-border-radius: var(--border-radius);
  --gallery-box-shadow: var(--medium-box-shadow);

  position: absolute;
  left: var(--left-right-inset);
  right: var(--left-right-inset);
  top: 0;
  height: var(--gallery-height);
  display: flex;
  flex-direction: column;
  padding: 10px 16px 16px;
  z-index: var(--z-09-gallery);
  border-bottom-left-radius: var(--gallery-border-radius);
  border-bottom-right-radius: var(--gallery-border-radius);
  background: white;
  box-shadow: var(--gallery-box-shadow);
  overflow: hidden;
  pointer-events: none;
  transform: translateY(calc(-1 * var(--gallery-height))) rotateX(25deg);
  transform-origin: 50% 0;
}

.gallery-loading,
.gallery-error {
  position: absolute;
  inset: 0;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  z-index: var(--z-03-gallery-message);

  a {
    color: inherit;
  }
}

body:not(.gallery-no-move-transition) .main-screen,
body:not(.gallery-no-move-transition) .gallery-panel,
body:not(.gallery-no-move-transition) .street-nameplate-container,
body:not(.gallery-no-move-transition) .street-section-sky {
  transition: transform 800ms;
}

body.gallery-visible .gallery-panel {
  transform: translateY(0);
  pointer-events: auto;
}
