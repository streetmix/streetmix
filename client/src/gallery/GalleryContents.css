.gallery-header {
  display: flex;
  flex-direction: row;
  flex: 0 0 40px;
  align-items: center;
  gap: 0.5em;

  .avatar {
    flex: 0 0 32px;
    height: 32px;
    margin-block-start: -1px;
  }
}

.gallery-label {
  flex-grow: 1;
  font-size: 1.5em;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.gallery-street-count {
  font-size: 1.5em;
  color: rgb(128 128 128);
  white-space: nowrap;
}

.gallery-streets-container {
  display: flex;
  flex-direction: row;
  gap: 1rem;
}

/* Applied by the wrapping element created by <Scrollable /> */
.streets-scrollable-container {
  position: relative;
  top: 1px;
  height: var(--gallery-height);
  min-width: 0; /* Allows content to overflow flex element */
  flex-grow: 1;

  button.scrollable {
    margin-top: 45px;
    width: 30px;

    &.scroll-left {
      left: -10px !important;
    }

    &.scroll-right {
      right: -10px !important;
    }
  }

  .streets {
    list-style: none;
    padding: 0;
    margin: 0;
    white-space: nowrap;
    height: calc(var(--gallery-height) + 20px); /* To cover the scroll bar */
    overflow-x: scroll;
  }
}

.gallery-user-buttons {
  display: flex;
  flex: 0 0 100px;
  flex-direction: column;
  height: 110px;
  margin-top: 6px;

  a.btn,
  button {
    padding-inline: 0.5em;
    display: flex !important;
    align-items: center !important;
    justify-content: center;
  }
}

.gallery-new-street {
  flex-grow: 1;
  font-weight: 550;
}

.gallery-copy-last-street {
  height: 40px;
  flex-basis: 40px;
  margin-top: 10px;
}
