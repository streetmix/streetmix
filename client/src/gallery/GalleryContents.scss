@use "sass:color";
@import "../../styles/variables";

.gallery-header {
  display: flex;
  flex-direction: row;
  flex: 0 0 40px;
  align-items: center;

  .avatar {
    flex: 0 0 32px;
    height: 32px;
    margin-top: -1px;
    margin-right: 0.5em;

    [dir="rtl"] & {
      margin-left: 0.5em;
      margin-right: 0;
    }
  }
}

.gallery-label {
  flex-grow: 1;
  font-size: 1.5em;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.gallery-street-count {
  font-size: 1.5em;
  color: rgb(128 128 128);
  white-space: nowrap;
}

.gallery-streets-container {
  display: flex;
  flex-direction: row;
}

// Applied by the wrapping element created by Scrollable.jsx component
.streets-scrollable-container {
  position: relative;
  top: 1px;
  height: $gallery-height;
  min-width: 0; // Allows content to overflow flex element
  flex-grow: 1;

  button.scrollable {
    margin-top: 45px;
    width: 30px;

    &.scroll-left {
      left: -10px !important;
    }

    &.scroll-right {
      right: -10px !important;
    }
  }

  .streets {
    list-style: none;
    padding: 0;
    margin: 0;
    white-space: nowrap;
    height: $gallery-height + 20px; // To cover the scroll bar
    overflow-x: scroll;
  }
}

.gallery-user-buttons {
  display: flex;
  flex: 0 0 100px;
  flex-direction: column;
  height: 110px;
  margin-top: 6px;
  margin-right: 16px;

  a.btn,
  button {
    padding-left: 0.5em;
    padding-right: 0.5em;
    display: flex !important;
    align-items: center !important;
    justify-content: center;
  }

  [dir="rtl"] & {
    margin-left: 16px;
    margin-right: 0;
  }
}

.gallery-new-street {
  flex-grow: 1;
  font-weight: 550;
}

.gallery-copy-last-street {
  height: 40px;
  flex-basis: 40px;
  margin-top: 10px;
}
