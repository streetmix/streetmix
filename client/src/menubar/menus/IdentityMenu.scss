@use "sass:color";
@import "../../../styles/variables";

.identity-menu {
  white-space: nowrap;
}

.identity-section {
  border-bottom: 1px solid $colour-turquoise-200;
  padding: 0.8em;
  min-width: 150px;
  max-width: 300px;
}

.identity-avatar-name {
  display: flex;
  flex-direction: row;
}

.identity-avatar-name-left {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  margin-right: 0.5em;

  [dir="rtl"] & {
    margin-left: 0.5em;
    margin-right: 0;
  }

  .avatar {
    width: 36px;
    height: 36px;
    margin: 0;
  }
}

.identity-avatar-name-right {
  display: flex;
  flex-direction: column;
  justify-content: center;
  line-height: 1.2;
  white-space: normal;
}

.identity-avatar-display-name {
  font-weight: 550;
  color: $colour-turquoise-700;
}

.identity-roles {
  margin-top: 0.5em;
  margin-bottom: -0.5em;

  ul {
    margin: 0;
    padding: 0;
    display: flex;
    flex-wrap: wrap;
  }

  ul li {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border-radius: $border-radius;
    padding: 0.3em 0.6em;
    margin-right: 0.35em;
    margin-bottom: 0.35em;
    font-size: 0.9em;
    font-weight: 550;

    [dir="rtl"] & {
      margin-left: 0.35em;
      margin-right: 0;
    }
  }

  .role-badge-generic {
    background-color: $colour-midnight-100;
    color: $colour-midnight-600;
  }

  .role-badge-subscriber {
    background-color: $colour-turquoise-200;
    color: $colour-turquoise-700;

    img {
      height: 1em;
      margin-right: 0.35em;

      [dir="rtl"] & {
        margin-left: 0.35em;
        margin-right: auto;
      }
    }
  }

  .role-badge-admin {
    background-color: $alert-border-colour;
    color: color.scale($alert-border-colour, $lightness: -75%);
  }
}
