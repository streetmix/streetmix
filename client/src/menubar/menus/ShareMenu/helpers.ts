import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
