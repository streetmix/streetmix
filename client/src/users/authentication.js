import Cookies from 'js-cookie'
import { jwtDecode } from 'jwt-decode'
import * as Sentry from '@sentry/browser'

import USER_ROLES from '../../../app/data/user_roles.json'
import { app } from '../preinit/app_settings'
import { showError, ERRORS } from '../app/errors'
import { MODES, processMode, getMode, setMode } from '../app/mode'
import { generateFlagOverrides, applyFlagOverrides } from '../app/flag_utils'
import { formatMessage } from '../locales/locale'
import { setPromoteStreet } from '../streets/remix'
import { fetchStreetFromServer, createNewStreetOnServer } from '../streets/xhr'
import store from '../store'
import { updateSettings } from '../store/slices/settings'
import { setSignInData, clearSignInData } from '../store/slices/user'
import { showDialog } from '../store/slices/dialogs'
import { updateStreetIdMetadata } from '../store/slices/street'
import { addToast } from '../store/slices/toasts'
import { getUser, deleteUserLoginToken } from '../util/api'
import { loadSettings } from './settings'

const USER_ID_COOKIE = 'user_id'
const SIGN_IN_TOKEN_COOKIE = 'login_token'
const REFRESH_TOKEN_COOKIE = 'refresh_token'
const LOCAL_STORAGE_SIGN_IN_ID = 'sign-in'

export function doSignIn () {
  store.dispatch(showDialog('SIGN_IN'))
}

export function getSignInData () {
  return store.getState().user.signInData || {}
}

export function isSignedIn () {
  return store.getState().user.signedIn
}

/**
 * Clears sign in data on the client side. Use this when authentication
 * data has become corrupted or expired on the client side and needs to be reset.
 * Do not use this to sign out a user. For that, use signOut(), which ensures
 * that sign out data is also sent to the server.
 */
function clearAllClientSignInData () {
  store.dispatch(clearSignInData())
  window.localStorage.removeItem(LOCAL_STORAGE_SIGN_IN_ID)
  removeSignInCookies()
}

export function onStorageChange () {
  if (isSignedIn() && !window.localStorage[LOCAL_STORAGE_SIGN_IN_ID]) {
    setMode(MODES.FORCE_RELOAD_SIGN_OUT)
    processMode()
  } else if (!isSignedIn() && window.localStorage[LOCAL_STORAGE_SIGN_IN_ID]) {
    setMode(MODES.FORCE_RELOAD_SIGN_IN)
    processMode()
  }
}

function saveSignInDataLocally () {
  const signInData = getSignInData()
  if (signInData) {
    window.localStorage.setItem(
      LOCAL_STORAGE_SIGN_IN_ID,
      JSON.stringify(signInData)
    )
  } else {
    window.localStorage.removeItem(LOCAL_STORAGE_SIGN_IN_ID)
  }
}

function removeSignInCookies () {
  Cookies.remove(SIGN_IN_TOKEN_COOKIE)
  Cookies.remove(REFRESH_TOKEN_COOKIE)
  Cookies.remove(USER_ID_COOKIE)
}

export async function loadSignIn () {
  const signInCookie = Cookies.get(SIGN_IN_TOKEN_COOKIE)
  const refreshCookie = Cookies.get(REFRESH_TOKEN_COOKIE)
  const userIdCookie = Cookies.get(USER_ID_COOKIE)

  if (signInCookie && userIdCookie && refreshCookie) {
    store.dispatch(
      setSignInData({
        token: signInCookie,
        refreshToken: refreshCookie,
        userId: userIdCookie
      })
    )

    saveSignInDataLocally()
  } else if (window.localStorage[LOCAL_STORAGE_SIGN_IN_ID]) {
    // old login data is in localstorage but we don't have the cookies we need
    clearAllClientSignInData()
    setMode(MODES.AUTH_EXPIRED)
    processMode()
    return true
  }

  const signInData = getSignInData()

  // Check if token is valid
  // Note that jwtDecode does not actually validate tokens. We will need
  // another library to do that. For now, we use the try/catch to see if
  // throws an error. If so, we log and report this error so we can see
  // why tokens are invalid, then we force clearing all data so that user
  // can reset and start over.
  try {
    if (signInData.token) {
      jwtDecode(signInData.token)
    }
  } catch (error) {
    Sentry.captureMessage('Error parsing jwt token ', signInData?.token)
    clearAllClientSignInData()
    setMode(MODES.AUTH_EXPIRED)
    processMode()
    return true
  }

  const storage = JSON.parse(window.localStorage.getItem('flags'))
  const sessionOverrides = generateFlagOverrides(storage, 'session')

  let flagOverrides = []

  if (signInData && signInData.token && signInData.userId) {
    const decoded = jwtDecode(signInData.token)

    // Check if token has expired
    const currentDate = new Date()
    const expDate = new Date(decoded.exp * 1000)

    // Set the expiration date to one day early so that tokens
    // don't wait till the last moment to refresh
    expDate.setDate(expDate.getDate() - 1)

    if (currentDate >= expDate) {
      await refreshLoginToken(signInData.refreshToken)
    }
    flagOverrides = await fetchSignInDetails(signInData.userId)
  } else {
    store.dispatch(clearSignInData())
  }

  if (!flagOverrides) {
    flagOverrides = []
  }
  applyFlagOverrides(store.getState().flags, ...flagOverrides, sessionOverrides)

  _signInLoaded()

  return true
}

/**
 *
 * @param {String} refreshToken
 * @returns {Object}
 */
async function refreshLoginToken (refreshToken) {
  const requestBody = JSON.stringify({ token: refreshToken })
  try {
    const response = await window.fetch('/services/auth/refresh-login-token', {
      method: 'post',
      headers: {
        'Content-Type': 'application/json'
      },
      body: requestBody
    })

    if (!response.ok) {
      throw response
    }
  } catch (error) {
    errorRefreshLoginToken(error)
  }
}

function errorRefreshLoginToken (data) {
  if (data.status === 401) {
    signOut(true)

    showError(ERRORS.SIGN_IN_401, true)
    return
  } else if (data.status === 503) {
    showError(ERRORS.SIGN_IN_SERVER_FAILURE, true)
    return
  }

  // Fail silently
  store.dispatch(clearSignInData())
}

/**
 *
 * @param {String} userId
 * @returns {Array}
 */
async function fetchSignInDetails (userId) {
  try {
    // TODO: See if it's possible to use RTK Query's implementation of getUser
    // because that will cache user details.
    const response = await getUser(userId)

    if (response.status !== 200) {
      throw response
    }

    const { flags, roles = [] } = response.data

    const flagOverrides = [
      // all role flag overrides
      ...roles.map((key) =>
        generateFlagOverrides(USER_ROLES[key].flags, `role:${key}`)
      ),
      // user flag overrides
      generateFlagOverrides(flags, 'user')
    ]

    receiveSignInDetails(response.data)
    return flagOverrides
  } catch (error) {
    errorReceiveSignInDetails(error)
  }
}

function receiveSignInDetails (details) {
  const signInData = {
    ...getSignInData(),
    details
  }
  store.dispatch(setSignInData(signInData))
  saveSignInDataLocally()
}

function errorReceiveSignInDetails (data) {
  if (data.status === 401) {
    signOut(true)

    // showError(ERRORS.SIGN_IN_401, true)
    // TODO: Check to make sure that this is the correct place to display
    // this. Currently, this will display for all 401 errors, not just
    // when a valid user who was previously signed in has been signed out.
    store.dispatch(
      addToast({
        message: formatMessage(
          'error.auth-expired',
          'We automatically signed you out due to inactivity. Please sign in again.'
        ),
        component: 'TOAST_SIGN_IN',
        duration: Infinity
      })
    )

    return
  } else if (data.status === 503) {
    showError(ERRORS.SIGN_IN_SERVER_FAILURE, true)
    return
  }

  // Fail silently
  store.dispatch(clearSignInData())
}

export function onSignOutClick (event) {
  signOut(false)

  if (event) {
    event.preventDefault()
  }
}

function signOut (quiet) {
  const signInData = getSignInData()
  store.dispatch(
    updateSettings({
      lastStreetId: null,
      lastStreetNamespacedId: null,
      lastStreetCreatorId: null
    })
  )

  sendSignOutToServer(signInData.userId, quiet)
}

function sendSignOutToServer (userId, quiet) {
  return deleteUserLoginToken(userId)
    .then((response) => {
      if (!quiet) {
        receiveSignOutConfirmationFromServer()
      }
    })
    .catch(errorReceiveSignOutConfirmationFromServer)
    .finally(() => {
      removeSignInCookies()
      window.localStorage.removeItem(LOCAL_STORAGE_SIGN_IN_ID)
    })
}

function receiveSignOutConfirmationFromServer () {
  setMode(MODES.SIGN_OUT)
  processMode()
}

function errorReceiveSignOutConfirmationFromServer () {
  setMode(MODES.SIGN_OUT)
  processMode()
}

function _signInLoaded () {
  loadSettings()
  const street = store.getState().street
  let mode = getMode()

  const surveyStreetId = Cookies.get('last_survey_url')

  // hack to return user to the survey street after signing in
  if (surveyStreetId) {
    Cookies.remove('last_survey_url')
    if (mode === MODES.JUST_SIGNED_IN) {
      window.location = surveyStreetId
    }
  }

  if (
    mode === MODES.CONTINUE ||
    mode === MODES.JUST_SIGNED_IN ||
    mode === MODES.USER_GALLERY ||
    mode === MODES.SURVEY_FINISHED ||
    mode === MODES.GLOBAL_GALLERY
  ) {
    const settings = store.getState().settings

    if (settings.lastStreetId) {
      store.dispatch(
        updateStreetIdMetadata({
          creatorId: settings.lastStreetCreatorId,
          id: settings.lastStreetId,
          namespacedId: settings.lastStreetNamespacedId
        })
      )

      if (mode === MODES.JUST_SIGNED_IN && !street.creatorId) {
        setPromoteStreet(true)
      }

      if (mode === MODES.SURVEY_FINISHED) {
        store.dispatch(
          addToast({
            message: formatMessage(
              'error.survey-finished',
              'Survey complete. Congratulations and thank you!'
            ),
            duration: Infinity
          })
        )
        setMode(MODES.CONTINUE)
      }
      if (mode === MODES.JUST_SIGNED_IN) {
        setMode(MODES.CONTINUE)
      }
    } else {
      setMode(MODES.NEW_STREET)
    }
  }
  mode = getMode()

  switch (mode) {
    case MODES.EXISTING_STREET:
    case MODES.CONTINUE:
    case MODES.USER_GALLERY:
    case MODES.GLOBAL_GALLERY:
      fetchStreetFromServer()
      break
    case MODES.NEW_STREET:
    case MODES.NEW_STREET_COPY_LAST:
      if (app.readOnly) {
        showError(ERRORS.CANNOT_CREATE_NEW_STREET_ON_PHONE, true)
      } else {
        createNewStreetOnServer()
      }
      break
  }
}
