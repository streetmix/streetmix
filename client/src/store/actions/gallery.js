import { createAsyncThunk } from '@reduxjs/toolkit'
import { hideControls } from '../../segments/resizing'
import { fetchGalleryData } from '../../gallery/index'
import { updatePageUrl } from '../../app/page_url'
import { showError, ERRORS } from '../../app/errors'
import { MODES, getMode, setMode } from '../../app/mode'
import { onWindowFocus } from '../../app/focus'

export const openGallery = createAsyncThunk(
  'gallery/openGallery',
  async ({ userId, instant = false }, { dispatch, rejectWithValue }) => {
    updatePageUrl(true, userId)
    hideControls()

    // TODO: Handle modes better.
    if (
      getMode() === MODES.USER_GALLERY ||
      getMode() === MODES.GLOBAL_GALLERY
    ) {
      // Prevents showing old street before the proper street loads
      showError(ERRORS.NO_STREET, false)
    }

    // Fetch data and catch errors if fetch goes wrong
    try {
      const streets = await fetchGalleryData(userId)
      return { streets }
    } catch (error) {
      // If the server error is 404, special rejection value
      // to display a "not-found" screen without the gallery
      if (error.response?.status === 404) {
        return rejectWithValue({
          killGallery: true
        })
      }

      // Re-throw original error for normal rejection
      throw error
    }
  }
)

export const closeGallery = createAsyncThunk(
  'gallery/closeGallery',
  ({ instant = false } = {}, { dispatch, getState }) => {
    const state = getState()

    if (!state.errors.abortEverything) {
      updatePageUrl()
    }

    onWindowFocus()
    setMode(MODES.CONTINUE)

    return { instant }
  },
  {
    condition: (arg, { getState }) => {
      const { visible } = getState().gallery

      // If gallery isn't visible, don't dispatch again
      if (!visible) {
        return false
      }
    }
  }
)
