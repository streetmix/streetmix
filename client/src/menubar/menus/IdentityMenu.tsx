import React, { useCallback } from 'react'
import { FormattedMessage } from 'react-intl'
import { StarIcon, MixerHorizontalIcon, ExitIcon } from '@radix-ui/react-icons'
import USER_ROLES from '../../../../app/data/user_roles.json'
import { onSignOutClick } from '../../users/authentication'
import Avatar from '../../users/Avatar'
import { useSelector, useDispatch } from '../../store/hooks'
import { openGallery } from '../../store/actions/gallery'
import { showDialog } from '../../store/slices/dialogs'
import streetmixPlusIcon from '../../ui/icons/streetmix-plus.svg'
import Menu, { type MenuProps } from './Menu'
import './IdentityMenu.scss'

function IdentityMenu (props: MenuProps): React.ReactElement {
  const user = useSelector((state) => state.user.signInData?.details)
  const isSubscriber = useSelector(
    (state) => state.user.signedIn && state.user.isSubscriber
  )
  const offline = useSelector((state) => state.system.offline)
  const dispatch = useDispatch()
  const handleClickMyStreets = useCallback(
    (event: React.MouseEvent) => {
      event.preventDefault()
      void dispatch(openGallery({ userId: user.id }))
    },
    [user?.id, dispatch]
  )

  const isAdmin: boolean =
    user?.roles?.includes(USER_ROLES.ADMIN.value) ?? false
  const myStreetsLink = user?.id !== undefined ? `/${user.id}` : ''

  return (
    <Menu {...props} className="identity-menu">
      {!offline && (
        <>
          <div className="identity-section">
            <div className="identity-avatar-name">
              <div className="identity-avatar-name-left">
                <Avatar userId={user?.id} />
              </div>
              <div className="identity-avatar-name-right">
                <div className="identity-avatar-display-name">
                  {user?.displayName ?? user?.id}
                </div>
                {user?.displayName !== undefined && (
                  <div className="menu-item-subtext">{user.id}</div>
                )}
              </div>
            </div>
            <div className="identity-roles">
              <ul>
                {isSubscriber && (
                  <li className="role-badge-subscriber">
                    <img
                      className="subscriber-icon"
                      src={streetmixPlusIcon}
                      alt=""
                    />
                    Streetmix+&lrm;
                  </li>
                )}
                {isAdmin && <li className="role-badge-admin">Admin</li>}
                {/* <li className="role-badge-generic">
                  Beta tester
                </li>
                <li className="role-badge-generic">
                  Translator
                </li> */}
              </ul>
            </div>
          </div>
          <a href={myStreetsLink} onClick={handleClickMyStreets}>
            <StarIcon className="menu-item-icon-radix" />
            <FormattedMessage
              id="menu.item.my-streets"
              defaultMessage="My streets"
            />
          </a>
        </>
      )}
      <a onClick={() => dispatch(showDialog('SETTINGS'))}>
        <MixerHorizontalIcon className="menu-item-icon-radix" />
        <FormattedMessage id="menu.item.settings" defaultMessage="Settings" />
      </a>
      <a className="menu-item" onClick={onSignOutClick}>
        <ExitIcon className="menu-item-icon-radix" />
        <FormattedMessage id="menu.item.sign-out" defaultMessage="Sign out" />
      </a>
    </Menu>
  )
}

export default IdentityMenu
