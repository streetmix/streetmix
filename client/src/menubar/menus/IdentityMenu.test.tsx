import React from 'react'
import { vi } from 'vitest'
import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'

import { render } from '~/test/helpers/render'
import { openGallery } from '~/src/store/actions/gallery'
import { onSignOutClick } from '~/src/users/authentication'
import IdentityMenu from './IdentityMenu'

vi.mock('../../store/actions/gallery', () => ({
  openGallery: vi.fn((id) => ({ type: 'MOCK_ACTION' }))
}))
vi.mock('../../users/authentication', () => ({
  onSignOutClick: vi.fn()
}))

describe('IdentityMenu', () => {
  const initialState = {
    user: {
      signInData: {
        details: {
          id: 'foo',
          displayName: 'bar'
        }
      }
    }
  }

  it('renders', () => {
    const { asFragment } = render(<IdentityMenu isActive={true} />, {
      initialState
    })
    expect(asFragment()).toMatchSnapshot()
  })

  it('shows "My streets" when its link is clicked', async () => {
    render(<IdentityMenu isActive={true} />, { initialState })

    await userEvent.click(screen.getByText('My streets'))

    expect(openGallery).toBeCalledTimes(1)
    expect(openGallery).toBeCalledWith({ userId: 'foo' })
  })

  it('signs the user out when its link is clicked', async () => {
    render(<IdentityMenu isActive={true} />, { initialState })

    await userEvent.click(screen.getByText('Sign out'))

    expect(onSignOutClick).toBeCalledTimes(1)
  })
})
