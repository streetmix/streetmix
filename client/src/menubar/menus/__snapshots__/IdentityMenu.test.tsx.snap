// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IdentityMenu > renders 1`] = `
<DocumentFragment>
  <div
    class="menu identity-menu"
  >
    <div
      class="identity-section"
    >
      <div
        class="identity-avatar-name"
      >
        <div
          class="identity-avatar-name-left"
        >
          <span
            class="avatar"
          >
            <img
              alt="foo"
            />
          </span>
        </div>
        <div
          class="identity-avatar-name-right"
        >
          <div
            class="identity-avatar-display-name"
          >
            bar
          </div>
          <div
            class="menu-item-subtext"
          >
            foo
          </div>
        </div>
      </div>
      <div
        class="identity-roles"
      >
        <ul />
      </div>
    </div>
    <div
      class="menu-separator"
    />
    <a
      href="/foo"
    >
      <svg
        class="menu-item-icon"
        data-icon="star"
        data-icon-source="test"
      />
      My streets
    </a>
    <a>
      <svg
        class="menu-item-icon"
        data-icon="settings"
        data-icon-source="test"
      />
      Settings
    </a>
    <div
      class="menu-separator"
    />
    <a
      class="menu-item menu-sign-out"
    >
      <svg
        class="menu-item-icon"
        data-icon="sign-out"
        data-icon-source="test"
      />
      Sign out
    </a>
  </div>
</DocumentFragment>
`;
